### PR TITLE
Add conflicts-with constraint

### PR DESCRIPTION
**Pull request by darecoder1999**
Implemented conflicts-with scrutinizer as part of #2354

### CODE REVIEW

File: extensions/wikidata/tests/src/org/openrefine/wikidata/qa/MockConstraintFetcher.java [hunk 1]
```diff
@@ -59,6 +60,11 @@
     
     public static PropertyIdValue propertyOnlyPid = Datamodel.makeWikidataPropertyIdValue("P372");
 
+    public static PropertyIdValue conflictsWithPid = Datamodel.makeWikidataPropertyIdValue("P50");
+    public static PropertyIdValue pidConflictingStatement = Datamodel.makeWikidataPropertyIdValue("P31");
+    public static Value conflictingStatementValue = Datamodel.makeWikidataItemIdValue("P31");
```
Comment (wetneb) [inline extensions/wikidata/tests/src/org/openrefine/wikidata/qa/MockConstraintFetcher.java:65]: `Datamodel.makeWikidataItemIdValue("P31")` is incorrect, you are trying to create an `ItemIdValue` with a property identifier.

File: extensions/wikidata/module/langs/translation-en.json [hunk 2]
```diff
@@ -104,6 +104,8 @@
     "warnings-messages/remove-statements-with-invalid-format/body": "If these statements currently exist on Wikidata, this will solve constraint violations.",
     "warnings-messages/missing-inverse-statements/title": "Inverse statements missing for {added_property_entity}.",
     "warnings-messages/missing-inverse-statements/body": "Any {added_property_entity} statement such as the one from {source_entity} to {target_entity} should also be added in reverse with {inverse_property_entity}: in this case, {target_entity} {inverse_property_entity} {source_entity}.",
+    "warnings-messages/having-conflicts-with-statements/title": "{property_entity} is having conflicts with statement {added_property_entity}.",
```
Comment (wetneb) [inline extensions/wikidata/module/langs/translation-en.json:107]: I do not think the "is having" is necessary here: `"{property_entity} conflicts with {added_property_entity}"` is simpler.

File: extensions/wikidata/src/org/openrefine/wikidata/qa/WikidataConstraintFetcher.java [hunk 3]
```diff
@@ -340,4 +333,47 @@ public boolean usableOnItems(PropertyIdValue pid) {
...
     }
+
+    /**
+     * Returns the list of PropertyIdValues of conflicting statements
+     *
+     * @param pid:
+     *            the property having conflicts-with constraint
+     * @return
+     */
+    public List<Value> getConflictsWithProperties (PropertyIdValue pid){
```
Comment (wetneb) [inline extensions/wikidata/src/org/openrefine/wikidata/qa/WikidataConstraintFetcher.java:344]: Style: no space between the method name and the opening parenthesis, one space between the closing parenthesis and the opening curly brace.

File: extensions/wikidata/src/org/openrefine/wikidata/qa/ConstraintFetcher.java [hunk 4]
```diff
@@ -134,4 +135,27 @@
...
+    List<Value> getConflictsWithProperties (PropertyIdValue pid);
+
+    /**
+     * Returns the list of all the conflicting values
+     *
+     * @param pid:
+     *            the property having conflicts-with constraint
+     * @return
+     */
+    List<Value> getItemwithConflicts (PropertyIdValue pid);
```
Comment (wetneb) [inline extensions/wikidata/src/org/openrefine/wikidata/qa/ConstraintFetcher.java:155]: Didn't we agree that we needed two parameters here: the property on which the constraint is defined, and the property it conflicts with?

Comment (darecoder1999) [reply extensions/wikidata/src/org/openrefine/wikidata/qa/ConstraintFetcher.java:155]: That's what I want to discuss with you, that If we are following wikibase quality constraints implementation so I guess they have used the same kind of implementation, they are fetching the item list at once and then checking that if the conflicting property is having a value from that list. https://github.com/wikimedia/mediawiki-extensions-WikibaseQualityConstraints/blob/master/src/ConstraintCheck/Checker/ConflictsWithChecker.php#L119-L139

File: extensions/wikidata/src/org/openrefine/wikidata/qa/WikidataConstraintFetcher.java (hunk 3 above)
Comment (darecoder1999) [reply extensions/wikidata/src/org/openrefine/wikidata/qa/WikidataConstraintFetcher.java:344]: Okay, I'll do the changes.

File: extensions/wikidata/src/org/openrefine/wikidata/qa/ConstraintFetcher.java (hunk 4 above)
Comment (wetneb) [reply extensions/wikidata/src/org/openrefine/wikidata/qa/ConstraintFetcher.java:155]: Their architecture is quite different from what we have here: yes they fetch items in a single go, but in this context they are already looking at a single constraint definition anyway. We should make sure examples like https://www.wikidata.org/wiki/Property:P4013 are handled correctly.

Comment (wetneb) [reply extensions/wikidata/src/org/openrefine/wikidata/qa/ConstraintFetcher.java:155]: Happy to discuss that at our next meeting on tuesday if that helps :)

File: extensions/wikidata/src/org/openrefine/wikidata/qa/ConstraintFetcher.java [hunk 5]
```diff
@@ -134,4 +135,27 @@
...
     boolean usableOnItems(PropertyIdValue pid);
+
+    /**
+     * Returns the list of PropertyIdValues of conflicting statements
+     *
+     * @param pid:
+     *            the property having conflicts-with constraint
+     * @return
+     */
+    List<Value> getConflictsWithProperties (PropertyIdValue pid);
```
[the remark this answers is not in the document]
Comment (darecoder1999) [reply extensions/wikidata/src/org/openrefine/wikidata/qa/ConstraintFetcher.java:146]: I did that so I can make use of findValues method else we have to create a new method doing nearly same thing.

[the remark this answers is not in the document]
Comment (wetneb) [reply extensions/wikidata/src/org/openrefine/wikidata/qa/ConstraintFetcher.java:146]: I think it is better to have a clean signature for an interface: you can convert the values with a map (from the streaming API).

File: extensions/wikidata/src/org/openrefine/wikidata/qa/WikidataConstraintFetcher.java [hunk 6]
```diff
@@ -340,4 +333,40 @@ public boolean usableOnItems(PropertyIdValue pid) {
...
+
+    /**
+     * Returns the list of PropertyIdValues of conflicting statements
+     *
+     * @param pid:
+     *            the property having conflicts-with constraint
+     * @return
+     */
+    public List<PropertyIdValue> getConflictsWithProperties(PropertyIdValue pid) {
+        List<SnakGroup> specs = getSingleConstraint(pid, CONFLICTS_WITH_CONSTRAINT_QID);
```
Comment (wetneb) [inline extensions/wikidata/src/org/openrefine/wikidata/qa/WikidataConstraintFetcher.java:345]: Because you are using `getSingleConstraint` here, this means that for properties with multiple conflicts-with constraints, you will only check the first one.

Comment (darecoder1999) [reply extensions/wikidata/src/org/openrefine/wikidata/qa/WikidataConstraintFetcher.java:345]: Okay, let me try the alternatives. Thanks for the review. :)

File: extensions/wikidata/src/org/openrefine/wikidata/qa/ConstraintFetcher.java (hunk 4 above)
Comment (darecoder1999) [reply extensions/wikidata/src/org/openrefine/wikidata/qa/ConstraintFetcher.java]: @wetneb: I have made the changes proposed, still finding an issue. When there are multiple issues to be reported for conflicts with constraint only the one is reported at once. I have checked that the raiseWarning method is returning true for all the possible issues but still the issues are not flagged in the issues tab. Please help. Thank you.

File: extensions/wikidata/src/org/openrefine/wikidata/qa/scrutinizers/ConflictsWithScrutinizer.java [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+public class ConflictsWithScrutinizer extends EditScrutinizer {
+
+    public static final String type = "having-conflicts-with-statements";
+
+    @Override
+    public void scrutinize(ItemUpdate update) {
+        Map<PropertyIdValue, Value> propertyIdValueValueMap = new HashMap<>();
+        for (Statement statement : update.getAddedStatements()){
+            PropertyIdValue pid = statement.getClaim().getMainSnak().getPropertyId();
+            Value value = statement.getClaim().getMainSnak().getValue();
```
[the remark this answers is not in the document]
Comment (darecoder1999) [reply extensions/wikidata/src/org/openrefine/wikidata/qa/scrutinizers/ConflictsWithScrutinizer.java:20]: In the case of NoValueSnak & SomeValueSnak, it will make a call to the method implemented in SnakImpl which will return null. So, I guess it will not create any issues. If it is an issue, please let me know, I'll use some other way to implement the functionality.

**Comment by darecoder1999** [comment]
@wetneb: Please review the changes made. I have removed all the bugs I have encountered until now. This one was great fun :)

File: extensions/wikidata/src/org/openrefine/wikidata/qa/WikidataConstraintFetcher.java [hunk 8]
```diff
@@ -337,4 +342,57 @@ public boolean usableOnItems(PropertyIdValue pid) {
...
+     */
+    public List<Value> getItemWithConflicts(PropertyIdValue pid, PropertyIdValue conflictingPid) {
+        List<Statement> statementList = getConstraintsByType(pid, CONFLICTS_WITH_CONSTRAINT_QID).collect(Collectors.toList());
+        List<Value> itemList = new ArrayList<>();
+        for (Statement statement : statementList) {
+            List<SnakGroup> specs = statement.getClaim().getQualifiers();
+            boolean flag = false;
+            for(SnakGroup group : specs) {
+                for (Snak snak : group.getSnaks()) {
+                    if (snak.getValue() instanceof PropertyIdValueImpl && snak.getValue().equals(conflictingPid)){
```
Comment (wetneb) [inline extensions/wikidata/src/org/openrefine/wikidata/qa/WikidataConstraintFetcher.java:384]: Two things: it is better to avoid referring to `PropertyIdValueImpl`, you can just use `PropertyIdValue` which is the interface. Also, this type check is actually redundant with the equality check afterwards. You should just be able to do `conflictingPid.equals(snak.getValue())` (since you know `conflictingPid` is not null and its equals method will only return true for instances of `PropertyIdValue`).

File: extensions/wikidata/src/org/openrefine/wikidata/qa/WikidataConstraintFetcher.java [hunk 9]
```diff
@@ -337,4 +342,57 @@ public boolean usableOnItems(PropertyIdValue pid) {
...
+                        flag = true;
+                    } else if (snak.getValue() instanceof PropertyIdValueImpl && !snak.getValue().equals(conflictingPid)) {
```
Comment (wetneb) [inline extensions/wikidata/src/org/openrefine/wikidata/qa/WikidataConstraintFetcher.java:386]: Same here.

File: extensions/wikidata/src/org/openrefine/wikidata/qa/scrutinizers/ConflictsWithScrutinizer.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (wetneb) [reply extensions/wikidata/src/org/openrefine/wikidata/qa/scrutinizers/ConflictsWithScrutinizer.java:20]: That is fine then! It would be worth adding a test case with a `NoValueSnak` to demonstrate that, perhaps?

File: extensions/wikidata/src/org/openrefine/wikidata/qa/WikidataConstraintFetcher.java (hunk 8 above)
Comment (darecoder1999) [reply extensions/wikidata/src/org/openrefine/wikidata/qa/WikidataConstraintFetcher.java:384]: Changing PropertyIdValueImpl to PropertyIdValue sounds good to me but removing that check will not be good I guess. Because I have used that as in Snaks we get both Property and Item so we have to update flag value only in case when we get Property as Snak. So, in the 1st if condition it will work but in 2nd else if part it will not work as expected.

Comment (wetneb) [reply extensions/wikidata/src/org/openrefine/wikidata/qa/WikidataConstraintFetcher.java:384]: Can you give me an example of a case where `snak.getValue().equals(conflictingPid)` evaluates to `true` but `snak.getValue() instanceof PropertyIdValue` does not?

Comment (darecoder1999) [reply extensions/wikidata/src/org/openrefine/wikidata/qa/WikidataConstraintFetcher.java:384]: As I mentioned, it will work for sure but in the 2nd else if condition `snak.getValue() instanceof PropertyIdValue && !snak.getValue().equals(conflictingPid)`  if we remove the check to check the instance class then there might be the case that it belongs to ItemIdValue and it will make get executed as this will be true in that case `!snak.getValue().equals(conflictingPid)`. 

As an alternative, if we really want to remove the instanceof check then one thing that comes to mind is we can place the last if condition at 2nd position. It will be better I guess?

Comment (wetneb) [reply extensions/wikidata/src/org/openrefine/wikidata/qa/WikidataConstraintFetcher.java:384]: If `snak.getValue()` is an `ItemIdValue` then `snak.getValue().equals(conflictingPid)` will evaluate to `false` - check the `equals` implementation of `ItemIdValue` :)

Comment (darecoder1999) [reply extensions/wikidata/src/org/openrefine/wikidata/qa/WikidataConstraintFetcher.java:384]: I feel we both are looking at the issue differently.

> If `snak.getValue()` is an `ItemIdValue` then `snak.getValue().equals(conflictingPid)` will evaluate to `false` - check the `equals` implementation of `ItemIdValue` :)

I do agree with your explanation, that's why I'm saying it. If it evaluates to false for other ItemIdValue the `!` at the starting `!snak.getValue().equals(conflictingPid)` will make it true and it will get executed which I don't want for now. I have tried to remove the instanceof comparison but it didn't worked as expected. If you are insisting I'll try it once again for sure.

Comment (wetneb) [reply extensions/wikidata/src/org/openrefine/wikidata/qa/WikidataConstraintFetcher.java]: Could you explain what you mean by "it didn't worked as expected"? Perhaps the best thing to do in this case is to put a breakpoint on this `if` and use the debugger to find out what is happening here exactly :)

File: extensions/wikidata/src/org/openrefine/wikidata/qa/WikidataConstraintFetcher.java (hunk 9 above)
Comment (wetneb) [reply extensions/wikidata/src/org/openrefine/wikidata/qa/WikidataConstraintFetcher.java]: Here, instead of relying on the `PropertyIdValue` type of the snak value, you should rather rely on the `snak.getProperty()`'s value.

File: extensions/wikidata/src/org/openrefine/wikidata/qa/WikidataConstraintFetcher.java (hunk 8 above)
Comment (darecoder1999) [reply extensions/wikidata/src/org/openrefine/wikidata/qa/WikidataConstraintFetcher.java]: Because I want the 2nd else if condition to execute only when the case is it is not and ItemIdValue, it is a PropertyIdValue but its value is not equal to the conflictingPid. But when I remove the instanceof check from it, the condition executes even for ItemIdValue as expected according to your explanation. 

> If `snak.getValue()` is an `ItemIdValue` then `snak.getValue().equals(conflictingPid)` will evaluate to `false` - check the `equals` implementation of `ItemIdValue` :)

But I want to it to execute for PropertyIdValue only when its value isn't equal to conflictingPid.

Comment (wetneb) [reply extensions/wikidata/src/org/openrefine/wikidata/qa/WikidataConstraintFetcher.java]: > But I want to it to execute for PropertyIdValue only when its value isn't equal to conflictingPid.

Yes but that is not how you should do it: you should look at the property of the snak, not the type of its values (because there could potentially be some other parameter which would also have the same type of values). So: remove all `instanceof` checks and use `snak.getProperty()` to know what sort of parameter you are handling.

Comment (darecoder1999) [reply extensions/wikidata/src/org/openrefine/wikidata/qa/WikidataConstraintFetcher.java]: I have removed the instanceof checks. Is it okay now?

Comment (wetneb) [reply extensions/wikidata/src/org/openrefine/wikidata/qa/WikidataConstraintFetcher.java]: It is much better! Now I guess we still have the issue that you are relying on the order of the qualifiers (you expect the property first, and then the items, but there is no reason why they should necessarily appear in this order).

Comment (darecoder1999) [reply extensions/wikidata/src/org/openrefine/wikidata/qa/WikidataConstraintFetcher.java]: I tried a lot of properties having conflicts-with constraint and observed this pattern, hence proceeded with this implementation. I won't say testes all of them but some of them. The item values don't seem to have anything that relates it to the conflicting Pid I guess. What else I should do? Any suggestions?

Comment (wetneb) [reply extensions/wikidata/src/org/openrefine/wikidata/qa/WikidataConstraintFetcher.java]: I would do as follows:
- create a variable to store the conflicting pid and a list of conflicting values
- loop over all qualifier snaks (as you are doing)
  + if the property of the snak is the one that specifies the conflicting pid, then set the conflicting pid to the qualifier value
  + if the property of the snak is the one that specifies the conflicting values, then add the value to the list of conflicting values

Comment (darecoder1999) [reply extensions/wikidata/src/org/openrefine/wikidata/qa/WikidataConstraintFetcher.java]: Hope, I correctly understand what you tried to explain. :)

File: extensions/wikidata/src/org/openrefine/wikidata/qa/ConstraintFetcher.java [hunk 10]
```diff
@@ -142,6 +144,7 @@
     boolean usableOnItems(PropertyIdValue pid);
 
     /**
+<<<<<<< HEAD
```
Comment (tfmorris) [inline extensions/wikidata/src/org/openrefine/wikidata/qa/ConstraintFetcher.java:147]: This looks like a missed merge conflict

File: extensions/wikidata/src/org/openrefine/wikidata/qa/WikidataConstraintFetcher.java [hunk 11]
```diff
@@ -28,9 +28,7 @@
 import org.wikidata.wdtk.datamodel.implementation.QuantityValueImpl;
 import org.wikidata.wdtk.datamodel.interfaces.*;
 
-import java.util.ArrayList;
-import java.util.List;
-import java.util.Set;
+import java.util.*;
```
Comment (tfmorris) [inline extensions/wikidata/src/org/openrefine/wikidata/qa/WikidataConstraintFetcher.java:31]: Did your IDE do this without you asking? Can you configure a higher limit for wild card imports and/or review the diff to make sure unnecessary changes aren't included?

File: extensions/wikidata/src/org/openrefine/wikidata/qa/WikidataConstraintFetcher.java [hunk 12]
```diff
@@ -94,7 +92,10 @@
     public static String ALLOWED_ENTITY_TYPES_QID = "Q52004125";
     public static String ALLOWED_ITEM_TYPE_QID = "Q29934200";
     public static String ALLOWED_ENTITY_TYPES_PID = "P2305";
-    
+
+    public static String CONFLICTS_WITH_CONSTRAINT_QID = "Q21502838";
+    public static String CONFLICTS_WITH_PROPERTY_PID = "P2306";
+    public static String ITEM_OF_PROPERTY_CONSTRAINT_PID = "P2305";
```
Comment (tfmorris) [inline extensions/wikidata/src/org/openrefine/wikidata/qa/WikidataConstraintFetcher.java:98]: This appears to duplicate `ALLOWED_ENTITY_TYPES_PID` just above.

File: extensions/wikidata/tests/src/org/openrefine/wikidata/qa/MockConstraintFetcher.java [hunk 13]
```diff
@@ -66,6 +69,11 @@
     public static QuantityValue minValuePid = Datamodel.makeQuantityValue(new BigDecimal(0));
     public static QuantityValue maxValuePid = Datamodel.makeQuantityValue(new BigDecimal(150));
 
+    public static PropertyIdValue conflictsWithPid = Datamodel.makeWikidataPropertyIdValue("P50");
+    public static PropertyIdValue pidConflictingStatement = Datamodel.makeWikidataPropertyIdValue("P31");
+    public static Value conflictingStatementValue = Datamodel.makeWikidataItemIdValue("Q5");
+    public static Value conflictsWithValue = Datamodel.makeWikidataItemIdValue("Q36322");
```
Comment (tfmorris) [inline extensions/wikidata/tests/src/org/openrefine/wikidata/qa/MockConstraintFetcher.java:75]: I recognize Q5, but the rest would have to look up. For the benefit of future readers, it might be nice to give them mnemonic variable names or to put the string values in constants.

File: extensions/wikidata/src/org/openrefine/wikidata/qa/ConstraintFetcher.java (hunk 10 above)
Comment (wetneb) [reply extensions/wikidata/src/org/openrefine/wikidata/qa/ConstraintFetcher.java:147]: whoops, thanks for catching that!

File: extensions/wikidata/src/org/openrefine/wikidata/qa/WikidataConstraintFetcher.java (hunk 12 above)
Comment (wetneb) [reply extensions/wikidata/src/org/openrefine/wikidata/qa/WikidataConstraintFetcher.java:98]: While these happen to have the same value for Wikidata, it might not be true of all Wikibase instances (these constants should become user-configurable once we add proper Wikibase support). That being said the name of the constant could probably be adapted to reflect that it is specifically for that constraint.

File: extensions/wikidata/src/org/openrefine/wikidata/qa/WikidataConstraintFetcher.java (hunk 11 above)
Comment (darecoder1999) [reply extensions/wikidata/src/org/openrefine/wikidata/qa/WikidataConstraintFetcher.java:31]: Yes, it got updated on its own by the IDE. Sorry, I'll take care of it next time onwards for sure. For now, I'll create a PR resolving all the flaws. Thanks for the review. :)
